Bottom: ed45d8943a8c3bb6fdd212d7eb5941fb6cf80f9c
Top:    53211529f3f7d06ef68f0461582bf40c1c70aa30
Author: Jason Wang <jasowang@redhat.com>
Date:   2015-01-27 00:30:48 -0500

virtio-net: disable_cb in start_xmit()

Disable tx interrupt during start_xmit(). Not sure its effect but it may
introduce lots of spurious tx interrupts.

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 2ab331f0..9127214 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -959,6 +959,8 @@ static netdev_tx_t start_xmit(struct sk_buff *skb, struct net_device *dev)
 	/* timestamp packet in software */
 	skb_tx_timestamp(skb);
 
+	virtqueue_disable_cb(sq->vq);
+
 	/* Try to transmit */
 	err = xmit_skb(sq, skb);
 
@@ -978,9 +980,14 @@ static netdev_tx_t start_xmit(struct sk_buff *skb, struct net_device *dev)
 	if (sq->vq->num_free < 2+MAX_SKB_FRAGS)
 		netif_stop_subqueue(dev, qnum);
 
-	if (kick || netif_xmit_stopped(txq))
+	if (kick || netif_xmit_stopped(txq)) {
 		virtqueue_kick(sq->vq);
-
+		if (!virtqueue_enable_cb_delayed(sq->vq) &&
+		    napi_schedule_prep(&sq->napi)) {
+			virtqueue_disable_cb(sq->vq);
+			__napi_schedule(&sq->napi);
+		}
+	}
 	return NETDEV_TX_OK;
 }
